controllers/drplacementcontrol.go

```diff
@@ -0,0 +1,1372 @@
+
+		return requeue
+	}
+
+	if d.needStatusUpdate || d.statusUpdateTimeElapsed() {
+		if err := d.updateDRPCStatus(); err != nil {
+			d.log.Error(err, "failed to update status")
+
+			return requeue
+		}
+	}
+
+	requeue = !done
+	d.log.Info("Completed processing placement", "requeue", requeue)
+
+	return requeue
+}
+
+func (d *DRPCInstance) processPlacement() (bool, error) {
+	d.log.Info("Process DRPC Placement", "DRAction", d.instance.Spec.Action)
+
+	switch d.instance.Spec.Action {
+	case rmn.ActionFailover:
+		return d.runFailover()
+	case rmn.ActionRelocate:
+		return d.runRelocate()
+	}
+
+	// Not a failover or a relocation.  Must be an initial deployment.
+	return d.runInitialDeployment()
+}
+
+func (d *DRPCInstance) runInitialDeployment() (bool, error) {
+	d.log.Info("Running initial deployment")
+
+	const done = true
+
+	homeCluster, homeClusterNamespace := d.getHomeCluster()
+
+	if homeCluster == "" {
+		err := fmt.Errorf("PreferredCluster not set and unable to find home cluster in DRPCPlacementRule (%v)",
+			d.drpcPlacementRule)
+		// needStatusUpdate is not set. Still better to capture the event to report later
+		rmnutil.ReportIfNotPresent(d.reconciler.eventRecorder, d.instance, corev1.EventTypeWarning,
+			rmnutil.EventReasonDeployFail, err.Error())
+
+		return !done, err
+	}
+
+	d.log.Info(fmt.Sprintf("Using homeCluster %s for initial deployment, uPlRule Decision %v",
+		homeCluster, d.userPlacementRule.Status.Decisions))
+
+	// Check if we already deployed in the homeCluster or elsewhere
+	clusterName, deployed := d.isDeployed(homeCluster)
+	if deployed && clusterName != homeCluster {
+		return done, nil
+	}
+
+	if deployed && d.isUserPlRuleUpdated(homeCluster) {
+		// If for whatever reason, the DRPC status is missing (i.e. DRPC could have been deleted mistakingly and
+		// recreated again), we should update it with whatever status we are at.
+		if d.getLastDRState() == rmn.DRState("") {
+			d.instance.Status.PreferredDecision = d.userPlacementRule.Status.Decisions[0]
+			d.setDRState(rmn.Deployed)
+		}
+
+		return done, nil
+	}
+
+	return d.startDeploying(homeCluster, homeClusterNamespace)
+}
+
+func (d *DRPCInstance) getHomeCluster() (string, string) {
+	// Check if the user wants to use the preferredCluster
+	homeCluster := ""
+	homeClusterNamespace := ""
+
+	if d.instance.Spec.PreferredCluster != "" {
+		homeCluster = d.instance.Spec.PreferredCluster
+		homeClusterNamespace = homeCluster
+	}
+
+	if homeCluster == "" && d.drpcPlacementRule != nil && len(d.drpcPlacementRule.Status.Decisions) != 0 {
+		homeCluster = d.drpcPlacementRule.Status.Decisions[0].ClusterName
+		homeClusterNamespace = d.drpcPlacementRule.Status.Decisions[0].ClusterNamespace
+	}
+
+	return homeCluster, homeClusterNamespace
+}
+
+// isDeployed check to see if the initial deployment is already complete to this
+// homeCluster or elsewhere
+func (d *DRPCInstance) isDeployed(homeCluster string) (string, bool) {
```

Comment:
As this is called only from initial deployment, detect the following:
- VRG should be reported as primary only on one cluster, error otherwise
- If VRG is secondary elsewhere, ensure its status reports secondary transition completed (healthy), else error out

Reply:
Rephrasing this comment:
- Ensure VRG is not Secondary in homeCluster
    - Currently we check if VRG is Primary on homeCluster, and then check peerClusters. It maybe Secondary on home cluster, because spec was modified to remove a relocate action, and we had just marked current homeCluster as Secondary
    - As it stands, the code would return deployed as `false` and empty clusterName starting a deployment to homeCluster where VRG is Secondary
- Further ensure that VRG is not present on peerClusters in any state, so that there are no cleanup actions to perform
    - Currently, if it is Primary on homeCluster we return, and so it maybe Secondary on a peerCluster and require a cleanup
    - We could cleanup at the end of deployment, but for now I think we will retain deploy only for initial deployment, and not reconcile feasible alternate specs that the user may change it to